dali/operators/generic/expand_dims.cc

```diff
@@ -0,0 +1,115 @@
+
+namespace dali {
+
+DALI_SCHEMA(ExpandDims)
+  .DocStr(R"code(Insert new dimension[s] of extent 1 and inserts new entries in "
+    "the layout (new_axis_names) at these indices in the layout.)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .InputDox(0, "data", "TensorList", "Data to be expanded")
+  .PassThrough({{0, 0}})
+  .AllowSequences()
+  .SupportVolumetric()
+  .AddOptionalArg<int>("axes", R"code(Indices where to put new dimensions of size 1.)code",
+    std::vector<int>(), true)
+  .AddOptionalArg("new_axis_names", R"code(Names of new dimensions in data layout.
+  
+Size of ``new_axis_names`` should be equal to ``axe``s size. 
+If argument won't be provided new dimensions will have layout '?')code", TensorLayout(""));
+
+template <typename Backend>
+ExpandDims<Backend>::ExpandDims(const OpSpec &spec)
+    : Reshape<Backend>(spec, typename Reshape<Backend>::BypassInit()) {
+  axes_ = spec.GetRepeatedArgument<int>("axes");
+  DALI_ENFORCE(spec.HasArgument("axes"), make_string("``axes`` argument should be provided."));
+  for (auto axis : axes_) {
+    DALI_ENFORCE(0 <= axis, make_string("Axis value can't be negative"));
+  }
+  std::sort(axes_.begin(), axes_.end());
+  DALI_ENFORCE(std::adjacent_find(axes_.begin(), axes_.end()) == axes_.end(),
+    make_string("Specified at least twice same index to add new dimension."));
+
+  new_axis_names_ = spec.GetArgument<TensorLayout>("new_axis_names");
+  if (!new_axis_names_.empty()) {
+    DALI_ENFORCE(new_axis_names_.size() == axes_.size(), make_string("Specified ", axes_.size(),
+      " new dimensions, but layout specify ", new_axis_names_.size(), " new names"));
+  }
+  this->use_src_dims_ = true;
+}
+
+template <typename Backend>
+bool ExpandDims<Backend>::SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) {
+  output_desc.resize(1);
+  this->SetOutputType(ws);
+
+  GenerateSrcDims(ws);
+  this->CalculateOutputShape(ws);
+
+  output_desc[0].type = *(this->output_type_);
+  output_desc[0].shape = this->output_shape_;
+  // return false, because we don't want the executor to allocate anything
+  // - this operator returns pointer to input memory
+  return false;
+}
+
+template <typename Backend>
+void ExpandDims<Backend>::GenerateSrcDims(const Workspace &ws) {
+  auto &in = ws.template InputRef<Backend>(0);
+  const auto &input_shape = in.shape();
+  int ndim = input_shape.sample_dim();
+  auto in_layout = in.GetLayout();
+  DALI_ENFORCE(in_layout.size() == ndim || in_layout.empty(),
+    make_string("Layout for data has size ",
+    in_layout.size(), " but data has ", ndim, " dimensions."));
+
+  TensorLayout out_layout;
+  size_t axes_ind = 0;
+  this->src_dims_.clear();
+  for (int d = 0; d < ndim; d++) {
+    if (axes_[axes_ind] == d) {
+      this->src_dims_.push_back(-1);
+      out_layout += new_axis_names_.empty() ? '?' : new_axis_names_[axes_ind];
+      axes_ind++;
+    }
+    out_layout += in_layout.empty() ? '?' : in_layout[d];
```

Comment:
If the input doesn't have a well defined layout, the output layout should also be empty.

The logic should be something like:
```
if (1) input has layout
    if (2) new_axis_names_ match the length of axes:
        output layout is a product of inserting new_axis_names at `axes`
    else
        either (3) reset layout to empty or (4) insert `?`
else
    (5) output layout is empty
```
(1) - We should assume that 0D input with empty layout has proper layout
(2) - This differs from `new_axis_names.empty()` in case when `axes` are also empty
(3) - This is consistent with (5) for 0D and >0D
(4) - This is inconsistent

Elaborating on inconsistency between (4) and (5):
A 0D has both specified and empty layout (layout is not a nullable property).
Now, if we insert `?`, a 0D tensor will _never_ be expanded to one with empty layout, even though it's very likely that the user wants it - instead, the tensor will get a (possibly invalid) layout with multiple `?` inserted.

Examples of the proposed logic:

ndim = 2
layout = `HW`
axes = [2]
axis_names = `C`
output layout `HWC`

ndim = 2
layout
axes = [2]
axis_names = 'C'
output: error - specifying axis names requires an input with a proper layuout

ndim = 2
layout = HW
axes = []
new_axis_names = ""
output layout = HW

ndim = 0
layout = ""
axes = [0, 1]
new_axis_names = "HW"
output layout = HW

ndim = 0
layout = ""
axes = [0, 1]
new_axis_names = ""
output layout = ""

ndim = 2
layout = "HW"
axes = [0]
new_axis_names = ""
output layout = ""